### DISCUSSION TITLE
Migrating from Cpp to Rust

### DISCUSSION THREAD

**Discussion by vdemcak** [stm32-rs/stm32wlxx-hal#356] [Q&A]
Hello,
So far, my firmware codebase has been written in C and CPP combined with HAL drivers provided by STM32, but since there are a lot of ongoing issues, I had this idea to give Rust a try and see if it could be a better alternative.

Now, here's the issue; I'm not exactly a Rust expert (especially when it comes to Rust embedded). And as soon as I started, I hit a bunch of roadblocks. The main one right now is configuring the clocks. In the C HAL, there's a handy function that lets me set up all the necessary parameters, but I was unable to find the Rust alternative to it.

Would anyone be able to help me? I'm attaching a snippet of the current clock config written in C.

Thanks in advance.

```c
RCC_OscInitTypeDef RCC_OscInitStruct = {0};
RCC_ClkInitTypeDef RCC_ClkInitStruct = {0};

// Configure LSE Drive Capability
HAL_PWR_EnableBkUpAccess();
__HAL_RCC_LSEDRIVE_CONFIG(RCC_LSEDRIVE_LOW);

// Configure the main internal regulator output voltage
__HAL_PWR_VOLTAGESCALING_CONFIG(PWR_REGULATOR_VOLTAGE_SCALE2);

// Initializes the CPU, AHB and APB busses clocks
RCC_OscInitStruct.OscillatorType = RCC_OSCILLATORTYPE_LSI|RCC_OSCILLATORTYPE_LSE|RCC_OSCILLATORTYPE_MSI;
RCC_OscInitStruct.LSEState = RCC_LSE_ON;
RCC_OscInitStruct.MSIState = RCC_MSI_ON;
RCC_OscInitStruct.MSIClockRange = RCC_MSIRANGE_4;

// Configure the SYSCLKSource, HCLK, PCLK1 and PCLK2 clocks dividers
RCC_ClkInitStruct.ClockType = RCC_CLOCKTYPE_HCLK3|RCC_CLOCKTYPE_HCLK|RCC_CLOCKTYPE_SYSCLK|RCC_CLOCKTYPE_PCLK1|RCC_CLOCKTYPE_PCLK2;
RCC_ClkInitStruct.SYSCLKSource = RCC_SYSCLKSOURCE_MSI;
RCC_ClkInitStruct.AHBCLKDivider = RCC_SYSCLK_DIV1;

// Ensure that MSI is wake-up system clock
__HAL_RCC_WAKEUPSTOP_CLK_CONFIG(RCC_STOP_WAKEUPCLOCK_MSI);
```

**Comment by newAM** [comment]
This HAL is a lot more bare-bones than the ST provided C-HAL at the moment.

For setting up clocks there may be equivalent functions in the RCC module: https://docs.rs/stm32wlxx-hal/latest/stm32wlxx_hal/rcc/index.html

Otherwise it's a matter of tracking down the registers the C-HAL uses and sequencing the same in the rust HAL via the peripheral access crate.  This is rather tedious because of the layers of abstraction in the C-HAl, generally I read the reference manual and write the register sequences from that instead.
